test/bilinearinterpolation.jl

```diff
@@ -0,0 +1,160 @@
+            y = DA([ysourcecpu[j] for i in 1:nsourcex, j in 1:nsourcey]),
+        )
+        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
+        targetmesh = (
+            x = DA([xtargetcpu[i] for i in 1:ntargetx, j in 1:ntargety]),
+            y = DA([ytargetcpu[j] for i in 1:ntargetx, j in 1:ntargety]),
+        )
+        fsource = testfunc.(sourcemesh.x, sourcemesh.y)
+        ftargetexact = testfunc.(targetmesh.x, targetmesh.y)
+        # allocate ftarget
+        ftarget = DA{FT}(undef, ntargetx, ntargety)
+        # use bilinear interpolation
+        interpolatebilinear!(ftarget, bilinear, fsource)
+        l∞error = maximum(abs.((ftarget .- ftargetexact)[:]))
+        @test l∞error ≤ toler
+    end
+    return (trial_constructor, trial_interpolator)
+end
+
+function test_multilevel(
+    ::Type{DA},
+    ::Type{FT},
+    (xmin, xmax),
+    (ymin, ymax),
+    (zmin, zmax),
+    (nsourcex, nsourcey),
+    (ntargetx, ntargety),
+    nlevels;
+    toler,
+) where {DA, FT}
+    @testset "bilinear interpolation on $nlevels levels with $FT, nsource = ($nsourcex, $nsourcey), ntarget = ($ntargetx, $ntargety)" begin
+        # build source and target mesh
+        xsource, xtarget = get_uniform_column_grids(
+            DA,
+            FT,
+            xmin,
+            xmax,
+            xmin,
+            xmax,
+            nsourcex,
+            ntargetx,
+        )
+        ysource, ytarget = get_uniform_column_grids(
+            DA,
+            FT,
+            ymin,
+            ymax,
+            ymin,
+            ymax,
+            nsourcey,
+            ntargety,
+        )
+        z, _ = get_uniform_column_grids(
+            DA,
+            FT,
+            zmin,
+            zmax,
+            zmin,
+            zmax,
+            nlevels,
+            nlevels,
+        )
+        bilinear = Bilinear(xsource, ysource, xtarget, ytarget)
+        # build fsource
+        xsourcecpu, ysourcecpu, zcpu = Array(xsource), Array(ysource), Array(z)
+        sourcemesh = (
+            x = DA([
+                xsourcecpu[j] for i in 1:nlevels, j in 1:nsourcex,
+                k in 1:nsourcey
+            ]),
+            y = DA([
+                ysourcecpu[k] for i in 1:nlevels, j in 1:nsourcex,
+                k in 1:nsourcey
+            ]),
+            z = DA([
+                zcpu[i] for i in 1:nlevels, j in 1:nsourcex, k in 1:nsourcey
+            ]),
+        )
+        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
+        targetmesh = (
+            x = DA([
+                xtargetcpu[j] for i in 1:nlevels, j in 1:ntargetx,
+                k in 1:ntargety
+            ]),
+            y = DA([
+                ytargetcpu[k] for i in 1:nlevels, j in 1:ntargetx,
+                k in 1:ntargety
+            ]),
+            z = DA([
+                zcpu[i] for i in 1:nlevels, j in 1:ntargetx, k in 1:ntargety
+            ]),
+        )
```

Comment:
```suggestion
        iter(x, nx, ny, nz) = [x[i] for i in 1:nx, j in 1:ny, k in 1:nz]
        sourcemesh = (
            x = DA(iter(xsourcecpu, nlevels, nsourcex, nsourcey)),
            x = DA(iter(ysourcecpu, nlevels, nsourcex, nsourcey)),
            x = DA(iter(zcpu, nlevels, nsourcex, nsourcey)),
        )
        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
        targetmesh = (
            x = DA(iter(xtargetcpu, nlevels, ntargetx, ntargety)),
            x = DA(iter(ytargetcpu, nlevels, ntargetx, ntargety)),
            x = DA(iter(zcpu, nlevels, ntargetx, ntargety)),
        )
```
Can we reduce some code duplication here?

Reply:
Or
```julia
        src_iter(x) = [x[i] for i in 1:nlevels, j in 1:nsourcex, k in 1:nsourcey]
        trg_iter(x) = [x[i] for i in 1:nlevels, j in 1:ntargetx, k in 1:ntargety]
        sourcemesh = (
            x = DA(src_iter(xsourcecpu)),
            x = DA(src_iter(ysourcecpu)),
            x = DA(src_iter(zcpu)),
        )
        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
        targetmesh = (
            x = DA(trg_iter(xtargetcpu)),
            x = DA(trg_iter(ytargetcpu)),
            x = DA(trg_iter(zcpu)),
        )
```

Reply:
They are not equivalent!